challenger/lnd.go

```diff
@@ -84,49 +89,61 @@ func (l *LndChallenger) Start() error {
 	addIndex := uint64(0)
 	settleIndex := uint64(0)
 
-	// Get a list of all existing invoices on startup and add them to our
-	// cache. We need to keep track of all invoices, even quite old ones to
-	// make sure tokens are valid. But to save space we only keep track of
-	// an invoice's state.
+	// Paginate through all existing invoices on startup and add them to our
+	// cache. We need to keep track of all invoices to ensure tokens are
+	// valid.
 	ctx := l.clientCtx()
-	invoiceResp, err := l.client.ListInvoices(
-		ctx, &lnrpc.ListInvoiceRequest{
-			NumMaxInvoices: math.MaxUint64,
-		},
-	)
-	if err != nil {
-		return err
-	}
-
-	// Advance our indices to the latest known one so we'll only receive
-	// updates for new invoices and/or newly settled invoices.
-	l.invoicesMtx.Lock()
-	for _, invoice := range invoiceResp.Invoices {
-		// Some invoices like AMP invoices may not have a payment hash
-		// populated.
-		if invoice.RHash == nil {
-			continue
+	indexOffset := uint64(0)
+	for {
+		invoiceResp, err := l.client.ListInvoices(
+			ctx, &lnrpc.ListInvoiceRequest{
+				IndexOffset:    indexOffset,
+				NumMaxInvoices: invoiceQueryPageSize,
+			},
+		)
+		if err != nil {
+			return err
 		}
 
-		if invoice.AddIndex > addIndex {
-			addIndex = invoice.AddIndex
-		}
-		if invoice.SettleIndex > settleIndex {
-			settleIndex = invoice.SettleIndex
-		}
-		hash, err := lntypes.MakeHash(invoice.RHash)
-		if err != nil {
-			l.invoicesMtx.Unlock()
-			return fmt.Errorf("error parsing invoice hash: %v", err)
+		// Lock the mutex to safely update the invoice states.
+		l.invoicesMtx.Lock()
+		for _, invoice := range invoiceResp.Invoices {
+			// Skip invoices that do not have a payment hash
+			// populated.
+			if invoice.RHash == nil {
+				continue
+			}
+
+			if invoice.AddIndex > addIndex {
+				addIndex = invoice.AddIndex
+			}
+			if invoice.SettleIndex > settleIndex {
+				settleIndex = invoice.SettleIndex
+			}
+			hash, err := lntypes.MakeHash(invoice.RHash)
+			if err != nil {
+				l.invoicesMtx.Unlock()
+				return fmt.Errorf("error parsing invoice "+
+					"hash: %v", err)
+			}
+
+			// Skip tracking the state of canceled or expired
+			// invoices.
+			if invoiceIrrelevant(invoice) {
+				continue
+			}
+			l.invoiceStates[hash] = invoice.State
 		}
+		l.invoicesMtx.Unlock()
 
-		// Don't track the state of canceled or expired invoices.
-		if invoiceIrrelevant(invoice) {
-			continue
+		// If there are no more invoices, stop pagination.
+		if len(invoiceResp.Invoices) < invoiceQueryPageSize {
```

Comment:
This seems to work with the current implementation of the invoice paginator that always returns exactly the maximum number of invoices. But on an index based query where we might have "holes" from deleted items, this would abort too early.
So with the risk of making one query too much, should we just check `len(invoiceResp.Invoices) == 0`?

Reply:
Yeah good idea! Done.